web_monitoring/diffing_server.py

```diff
@@ -115,8 +116,47 @@ def options(self):
 class DiffHandler(BaseHandler):
     # subclass must define `differs` attribute
 
+    # If query parameters repeat, take last one.
+    # Decode clean query parameters into unicode strings and cache the results.
+    @functools.lru_cache()
+    def decode_query_params(self):
+        query_params = {k: v[-1].decode() for k, v in
+                        self.request.arguments.items()}
+        return query_params
+
+    # Compute our own ETag header values.
+    def compute_etag(self):
+        # We're not actually hashing content for this, since that is expensive.
+        validation_bytes = str(
+            web_monitoring.__version__
+            + self.request.path
+            + str(self.decode_query_params())
+        ).encode('utf-8')
+
+        # Uses the "weak validation" directive since we don't guarantee that future
+        # responses for the same diff will be byte-for-byte identical.
+        etag = str('W/"' + web_monitoring.utils.hash_content(validation_bytes) + '"').encode('utf-8')
```

Comment:
Please use a format string for this. It’s more concise and readable.

```suggestion
        etag = f'W/"{web_monitoring.utils.hash_content(validation_bytes)}"'.encode('utf-8')
```

Reply:
Also, now that we have tests, I ran them without encoding this return value as bytes and it seems to work fine. Is there any reason we are doing it?

```suggestion
        etag = f'W/"{web_monitoring.utils.hash_content(validation_bytes)}"'
```

Reply:
No good reason at all. This is much nicer.